source/extensions/filters/http/cache/cache_headers_utils.cc

```diff
@@ -287,52 +303,40 @@ constexpr absl::string_view header_separator = "\n";
 constexpr absl::string_view in_value_separator = "\r";
 }; // namespace
 
-std::string VaryHeader::createVaryKey(const Http::HeaderMap::GetResult& vary_header,
-                                      const Http::RequestHeaderMap& entry_headers) {
-  if (vary_header.empty()) {
-    return "";
+absl::optional<std::string>
+VaryHeader::createVaryIdentifier(const absl::btree_set<absl::string_view>& vary_header_values,
+                                 const Http::RequestHeaderMap& request_headers) const {
+  std::string vary_identifier = "vary-id\n";
+  if (vary_header_values.empty()) {
+    return vary_identifier;
   }
 
-  // TODO(mattklein123): Support multiple vary headers and/or just make the vary header inline.
-  ASSERT(vary_header[0]->key() == "vary");
-
-  std::string vary_key = "vary-key\n";
-
-  for (const std::string& header : CacheHeadersUtils::parseCommaDelimitedList(vary_header)) {
-    // TODO(cbdm): Can add some bucketing logic here based on header. For example, we could
-    // normalize the values for accept-language by making all of {en-CA, en-GB, en-US} into
-    // "en". This way we would not need to store multiple versions of the same payload, and any
-    // of those values would find the payload in the requested language. Another example would be to
-    // bucket UserAgent values into android/ios/desktop; UserAgent::initializeFromHeaders tries to
-    // do that normalization and could be used as an inspiration for some bucketing configuration.
-    // The config should enable and control the bucketing wanted.
+  for (const absl::string_view& header : vary_header_values) {
+    if (header.empty()) {
+      continue;
+    }
+    if (!allowsHeader(header)) {
```

Comment:
Is this expected? Perhaps an assert is more appropriate here if that should never happen.

Reply:
I think it is expected. This is supposed to handle cases where the backend provides a value that our config does not permit varying.

Reply:
Perhaps mention in the comment that the vary list can change and headers cached by the backend can get stale in that case?

Reply:
ping re: comment

Reply:
Done